File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/IfThenElse.scala
package org.enso.compiler.core.ir
package expression

import org.enso.compiler.core.Implicits.{ShowPassData, ToStringHelper}
import org.enso.compiler.core.{IR, Identifier}
import org.enso.compiler.core.IR.{indentLevel, mkIndent}

import java.util.UUID

/** The Enso case expression. */
case class IfThenElse(
  cond: Expression,
  trueBranch: Expression,
  private val falseBranchOrNull: Expression,
  override val identifiedLocation: IdentifiedLocation,
  override val passData: MetadataStorage = new MetadataStorage()
) extends Expression
    with IRKind.Primitive
    with LazyDiagnosticStorage
    with LazyId {

  cond.getClass()
  trueBranch.getClass()

  def falseBranch(): Option[Expression] = Option(falseBranchOrNull)

  /** Creates a copy of `this`.
    *
    * @param cond   the expression whose value is being matched on
    * @param branches    the branches of the case expression
    * @param isNested    if true, the flag indicates that the expr represents a desugared nested case
    * @param location    the source location that the node corresponds to
    * @param passData    the pass metadata associated with this node
    * @param diagnostics compiler diagnostics for this node
    * @param id          the identifier for the new node
    * @return a copy of `this`, updated with the specified values
    */
  def copy(
    cond: Expression                     = cond,
    trueBranch: Expression               = trueBranch,
    falseBranchOrNull: Expression        = falseBranchOrNull,
    location: Option[IdentifiedLocation] = location,
    passData: MetadataStorage            = passData,
    diagnostics: DiagnosticStorage       = diagnostics,
    id: UUID @Identifier                 = id
  ): IfThenElse = {
    if (
      cond != this.cond
      || trueBranch != this.trueBranch
      || falseBranchOrNull != this.falseBranchOrNull
      || location != this.location
      || passData != this.passData
      || diagnostics != this.diagnostics
      || id != this.id
    ) {
      val res = new IfThenElse(
        cond,
        trueBranch,
        falseBranchOrNull,
        location.orNull,
        passData
      )
      res.diagnostics = diagnostics
      res.id          = id
      res
    } else this
  }

  /** @inheritdoc */
  override def mapExpressions(
    fn: java.util.function.Function[Expression, Expression]
  ): IfThenElse = {
    copy(
      cond       = fn(cond),
      trueBranch = fn(trueBranch),
      falseBranchOrNull =
        if (falseBranchOrNull ne null) fn(falseBranchOrNull) else null
    )
  }

  /** @inheritdoc */
  override def setLocation(location: Option[IdentifiedLocation]): IfThenElse =
    copy(location = location)

  /** @inheritdoc */
  override def duplicate(
    keepLocations: Boolean   = true,
    keepMetadata: Boolean    = true,
    keepDiagnostics: Boolean = true,
    keepIdentifiers: Boolean = false
  ): IfThenElse = {
    copy(
      cond = cond.duplicate(
        keepLocations,
        keepMetadata,
        keepDiagnostics,
        keepIdentifiers
      ),
      trueBranch = trueBranch.duplicate(
        keepLocations,
        keepMetadata,
        keepDiagnostics,
        keepIdentifiers
      ),
      falseBranchOrNull =
        if (falseBranchOrNull ne null)
          falseBranchOrNull.duplicate(
            keepLocations,
            keepMetadata,
            keepDiagnostics,
            keepIdentifiers
          )
        else null,
      location = if (keepLocations) location else None,
      passData =
        if (keepMetadata) passData.duplicate else new MetadataStorage(),
      diagnostics = if (keepDiagnostics) diagnosticsCopy else null,
      id          = if (keepIdentifiers) id else null
    )
  }

  /** String representation. */
  override def toString: String =
    s"""
         |IfThenElse(
         |cond = $cond,
         |trueBranch = $trueBranch,
         |falseBranch = ${falseBranch()},
         |location = $location,
         |passData = ${this.showPassData},
         |diagnostics = $diagnostics,
         |id = $id
         |)
         |""".toSingleLine

  /** @inheritdoc */
  override def children: List[IR] =
    List(cond, trueBranch) ++ falseBranch().toList

  /** @inheritdoc */
  override def showCode(indent: Int): String = {
    val newIndent = indent + indentLevel
    val headerStr =
      s"if ${cond.showCode(indent)} then\n${trueBranch.showCode(newIndent)}"
    val elseStr =
      if (falseBranchOrNull ne null)
        s"${mkIndent(indent)}else\n${falseBranchOrNull.showCode(newIndent)}"
      else ""
    s"$headerStr\n$elseStr"
  }

}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/expression/IfThenElse.java
package org.enso.compiler.core.ir.expression;

import org.enso.compiler.core.ir.Empty;
import org.enso.compiler.core.ir.Expression;
import org.enso.compiler.core.ir.IdentifiedLocation;
import org.enso.compiler.core.ir.MetadataStorage;
import org.enso.runtime.parser.dsl.GenerateFields;
import org.enso.runtime.parser.dsl.GenerateIR;
import org.enso.runtime.parser.dsl.IRChild;

/** Enso if/then(else) expression. */
@GenerateIR(interfaces = {Expression.class})
public final class IfThenElse extends IfThenElseGen {

  @GenerateFields
  public IfThenElse(
      @IRChild Expression condition,
      @IRChild Expression trueBranch,
      @IRChild(required = false) Expression falseBranchOrNull,
      IdentifiedLocation identifiedLocation,
      MetadataStorage passData) {
    super(condition, trueBranch, falseBranchOrNull, identifiedLocation, passData);
  }

  public Expression cond() {
    return condition();
  }

  public IfThenElse copy(Expression cond) {
    return builder(this).condition(cond).build();
  }

  public IfThenElse copy(Expression cond, Expression trueBranch, Expression falseBranchOrNull) {
    return builder(this)
        .condition(cond)
        .trueBranch(trueBranch)
        .falseBranchOrNull(falseBranchOrNull)
        .build();
  }

  public scala.Option<Expression> falseBranch() {
    return scala.Option.apply(falseBranchOrNull());
  }

  /**
   * Builds "else only" instance.
   *
   * @param value the else branch
   * @param where location of the value
   * @param meta the metadaa in the tree
   * @return insance of if/then/else
   * @see #isOnlyElse()
   */
  public static IfThenElse buildOnlyElse(
      Expression value, IdentifiedLocation where, MetadataStorage meta) {
    var emptyBlock = Empty.builder().build();
    return builder()
        .condition(emptyBlock)
        .trueBranch(emptyBlock)
        .falseBranchOrNull(value)
        .location(where)
        .passData(meta)
        .build();
  }

  /**
   * Checks whether this object represents "pending" else branch. Such a branch only makes sense if
   * merged to previous if statement.
   *
   * @return {@code true} if created by {@link #buildOnlyElse} factory
   * @see #isOnlyElse()
   */
  public boolean isOnlyElse() {
    return cond() instanceof Empty c && trueBranch() instanceof Empty t && c == t;
  }

  @Override
  public String showCode(int indent) {
    var newIndent = indent + indentLevel;
    var headerStr = "if " + cond().showCode(0) + " then\n" + trueBranch().showCode(newIndent);
    var elseStr =
        switch (falseBranchOrNull()) {
          case Expression f -> " ".repeat(indent) + "else\n" + f.showCode(newIndent);
          case null -> "";
        };
    return headerStr + "\n" + elseStr;
  }
}
